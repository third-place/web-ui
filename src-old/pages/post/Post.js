import { Button } from '@mui/material';
import React, { useContext, useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import MetaTags from 'react-meta-tags';
import { getPost } from '../../actions/post';
import { createReply, getReplies } from '../../actions/reply';
import Container from '../../components/Container';
import { imageBaseUrl } from '../../utils/config';
import Reply from './components/Reply';
import TextInput from '../../components/TextInput';
import { default as PostComponent } from '../../components/Post';
import Context from '../../utils/Context';

export default function Post() {
  const [post, setPost] = useState(null);
  const [replies, setReplies] = useState([]);
  const [reply, setReply] = useState('');
  const { posts, sessionToken, loggedInUser, isLoggedIn } = useContext(Context);
  const params = useParams();
  const navigate = useNavigate();

  const loadPost = async () => {
    const response = await getPost(sessionToken, params.uuid);
    const data = await response.json();
    setPost(data);
  };

  const loadReplies = async () => {
    const response = await getReplies(sessionToken, params.uuid);
    const data = await response.json();
    setReplies(data);
  };

  useEffect(() => {
    for (const p of posts) {
      if (p.uuid === params.uuid) {
        setPost(p);
        loadReplies();
        return;
      }
    }
    (async function() {
      await loadPost();
      await loadReplies();
    })();
  }, []);

  if (!post) {
    return <Container />;
  }

  const trySubmit = async (event) => {
    event.preventDefault();
    const response = await createReply(sessionToken, params.uuid, loggedInUser.uuid, reply);
    if (response.status === 201) {
      setReply("");
      await loadReplies();
    }
  };

  const title = `Read ${post.user.name || '@'+post.user.username}'s Post`;
  const description = post.text.split(".")[0];

  return (
    <Container>
      <MetaTags>
        <title>{title}</title>
        <meta
          property="og:title"
          content={title}
        />
        <meta
          property="og:url"
          content={`https://thirdplaceapp.com/p/${post.uuid}`}
        />
        { post.images && post.images.length > 0 && (
          <meta
            property="og:image"
            content={`${imageBaseUrl}/${post.images[0].s3_key}`}
          />
        )}
        <meta property="og:description" content={description} />
        <meta property="og:site_name" content="Third Place" />
        <meta property="og:type" content="article" />
        <meta property="article:published_time" content={post.created_at} />
      </MetaTags>
      <PostComponent post={post} onDelete={() => navigate("/")} showReply={false} />
      <h3>Replies</h3>
      { isLoggedIn && (
        <form onSubmit={trySubmit} style={{paddingBottom: 20}}>
          <TextInput
            label="Share your reply"
            variant="standard"
            onChangeValue={setReply}
            value={reply}
            multiline
            style={{width: 500}}
          />
          <Button type="submit">
            Reply
          </Button>
        </form>
      )}
      {replies.map((r) => (
        <Reply key={r.uuid} reply={r} />
      ))}
    </Container>
  )
}
