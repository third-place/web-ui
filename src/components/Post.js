import {
  Avatar,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Paper,
  Typography,
  Link, Card, IconButton,
} from '@mui/material';
import nl2br from 'react-nl2br';
import React, { useContext, useEffect, useState } from 'react';
import { Link as RouterLink, useNavigate } from 'react-router-dom';
import ReactMarkdown from 'react-markdown';
import ChatBubbleOutlineIcon from '@mui/icons-material/ChatBubbleOutline';
import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder';
import FavoriteIcon from '@mui/icons-material/Favorite';
import RepeatIcon from '@mui/icons-material/Repeat';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ExpandLessIcon from '@mui/icons-material/ExpandLess';
import LinkIcon from '@mui/icons-material/Link';
import TimeAgo from 'javascript-time-ago';
import en from 'javascript-time-ago/locale/en';
import { createPostLike, deletePostLike } from '../actions/like';
import { deletePost } from '../actions/post';
import { imageBaseUrl } from '../utils/config';
import Context from '../utils/Context';
import PostMenu from './PostMenu';

TimeAgo.addDefaultLocale(en);
const timeAgo = new TimeAgo('en-US');

export default function Post({
  post: {
    uuid,
    text,
    created_at,
    user: author,
    images,
    selfLiked,
    share,
  },
  onDelete,
  onUnlike,
  showReply,
  showShare,
  sharePostClick,
  compact,
}) {
  const [isDialogOpen, setIsDialogOpen] = useState(false);
  const [isSelfLiked, setIsSelfLiked] = useState(selfLiked);
  const [isExpanded, setIsExpanded] = useState(!compact);
  const [element, setElement] = useState(null);
  const [overflows, setOverflows] = useState(false);
  const [showCopiedMessage, setShowCopiedMessage] = useState(false);
  const { isLoggedIn, sessionToken, loggedInUser } = useContext(Context);
  const created = new Date(created_at);
  showShare = showShare === undefined || Boolean(showShare);
  onUnlike = onUnlike === undefined ? () => {} : onUnlike;
  const navigate = useNavigate();

  const tryDelete = async () => {
    await deletePost(sessionToken, uuid);
    onDelete();
    handleClose();
  };

  const handleClose = () => {
    setIsDialogOpen(false);
  };

  const tryLikePost = async () => {
    await createPostLike(sessionToken, uuid);
    setIsSelfLiked(true);
  };

  const tryUnlikePost = async () => {
    await deletePostLike(sessionToken, uuid);
    onUnlike();
    setIsSelfLiked(false);
  };

  useEffect(() => {
    if (element && compact) {
      setOverflows(element.scrollHeight > element.clientHeight);
    }
  }, [element]);

  const authorDisplayName = author.name ? author.name : "";
  const profilePic = author.profile_pic ? `${imageBaseUrl}/${author.profile_pic}` : '';

  const onClickExpand = () => {
    if (compact) {
      setIsExpanded(!isExpanded);
    }
  };

  const copyLinkToClipboard = (url) => {
    navigator.clipboard.writeText(url);
    setShowCopiedMessage(true);
    setTimeout(() => {
      setShowCopiedMessage(false);
    }, 2000);
  };

  return (
    <Card sx={{ p: 1, mb: 1 }}>
      { isLoggedIn && loggedInUser.uuid === author.uuid && (
        <PostMenu
          handleDelete={() => setIsDialogOpen(true)}
          handleEdit={() => navigate(`/p-edit/${uuid}`)}
        />
      )}
      <Avatar
        alt={authorDisplayName}
        src={profilePic}
        style={{ float: "left", marginRight: 10 }}
        sx={{ width: 56, height: 56 }}
      />
      <div style={{width: "90%", paddingLeft: 72}}>
        <Typography variant="h6">
          <Link component={RouterLink} to={`/u/${author.username}`}>
            <b>{author.name}</b>
            <span style={{color: "#f17887", margin: "0 20px"}}>@{author.username}</span>
            <span style={{fontSize: "smaller"}}>{timeAgo.format(created)}</span>
          </Link>
        </Typography>
        <div onClick={onClickExpand}>
          <div
            style={{
              maxHeight: isExpanded ? "inherit" : 190,
              cursor: overflows ? "pointer" : "auto",
            }}
            className="post"
            ref={(el) => setElement(el)}
          >
            <ReactMarkdown>
              {text}
            </ReactMarkdown>
          </div>
          <div style={{width: "100%", textAlign: "center"}}>
            { !isExpanded && overflows && (
              <IconButton>
                <ExpandMoreIcon cursor="pointer" />
              </IconButton>
            )}
            { isExpanded && overflows && (
              <IconButton>
                <ExpandLessIcon cursor="pointer" />
              </IconButton>
            )}
          </div>
        </div>
        { share && (
          <Paper
            sx={{p: 1, mb: 1, }}
            elevation={0}
            variant="outlined"
          >
            <Avatar
              alt={share.user.name}
              src={share.user.profile_pic ? `${imageBaseUrl}/${share.user.profile_pic}` : ''}
              style={{ float: "left", marginRight: 10 }}
            />
            <Link component={RouterLink} to={`/u/${share.user.username}`}>
              {share.user.name} @{share.user.username}
            </Link>
            <Typography color="text.secondary">
              {new Date(share.created_at).toLocaleString()}
            </Typography>
            <Typography variant="body2">
              {nl2br(share.text)}
            </Typography>
          </Paper>
        )}
        <div>
          {images && images.map((i) => (
            <Link component={RouterLink} to={`/i/${i.uuid}`} key={i.uuid}>
              <img src={`${imageBaseUrl}/${i.s3_key}`} className="post-gallery" alt="" />
            </Link>
          ))}
        </div>
      </div>
      { isLoggedIn && (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
          { showReply && (
            <IconButton aria-label="reply" component={RouterLink} to={`/p/${uuid}`}>
              <ChatBubbleOutlineIcon fontSize="small" />
            </IconButton>
          )}
          { !isSelfLiked && (
            <IconButton aria-label="like" onClick={tryLikePost}>
              <FavoriteBorderIcon fontSize="small" />
            </IconButton>
          )}
          { isSelfLiked && (
            <IconButton aria-label="unlike" onClick={tryUnlikePost}>
              <FavoriteIcon fontSize="small" />
            </IconButton>
          )}
          { showShare && (
            <IconButton aria-label="share" onClick={sharePostClick}>
              <RepeatIcon fontSize="small" />
            </IconButton>
          )}
          <IconButton aria-label="link" onClick={() => copyLinkToClipboard(`${window.location.hostname}/p/${uuid}`)}>
            <LinkIcon fontSize="small" />
            { showCopiedMessage && (
              <span style={{
                position: "absolute",
                left: 24,
                fontSize: "12pt",
                width: 120,
              }}>
                Link copied!
              </span>
            )}
          </IconButton>
        </div>
      )}
      <Dialog
        open={isDialogOpen}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">
          Delete this post?
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure? This cannot be undone!
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button onClick={tryDelete} autoFocus>
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </Card>
  );
}
