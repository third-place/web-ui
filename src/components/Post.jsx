import {
  Avatar,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Paper,
  Link, Card, IconButton, Typography,
} from '@mui/material';
import React, { useContext, useEffect, useState } from 'react';
import { Link as RouterLink, useNavigate } from 'react-router-dom';
import ReactMarkdown from 'react-markdown';
import ChatBubbleOutlineIcon from '@mui/icons-material/ChatBubbleOutline';
import FavoriteBorderIcon from '@mui/icons-material/FavoriteBorder';
import FavoriteIcon from '@mui/icons-material/Favorite';
import RepeatIcon from '@mui/icons-material/Repeat';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ExpandLessIcon from '@mui/icons-material/ExpandLess';
import LinkIcon from '@mui/icons-material/Link';
import { createPostLike, deletePostLike } from '../actions/like';
import { deletePost } from '../actions/post';
import { imageBaseUrl } from '../utils/config';
import Context from '../utils/Context';
import PostMenu from './PostMenu';
import TitleSlug from './TitleSlug';
import { timeAgo } from '../utils/timeAgo';

export default function Post({
  post: {
    uuid,
    text,
    created_at,
    user: author,
    images,
    selfLiked,
    share,
  },
  onDelete,
  onUnlike,
  showReply,
  showShare,
  sharePostClick,
  compact,
}) {
  const [isDialogOpen, setIsDialogOpen] = useState(false);
  const [isSelfLiked, setIsSelfLiked] = useState(selfLiked);
  const [isExpanded, setIsExpanded] = useState(!compact);
  const [element, setElement] = useState(null);
  const [overflows, setOverflows] = useState(false);
  const [showCopiedMessage, setShowCopiedMessage] = useState(false);
  const { isLoggedIn, sessionToken, loggedInUser } = useContext(Context);
  const created = new Date(created_at);
  showShare = showShare === undefined || Boolean(showShare);
  onUnlike = onUnlike === undefined ? () => {} : onUnlike;
  const navigate = useNavigate();

  const tryDelete = async () => {
    await deletePost(sessionToken, uuid);
    onDelete();
    handleClose();
  };

  const handleClose = () => {
    setIsDialogOpen(false);
  };

  const tryLikePost = async () => {
    await createPostLike(sessionToken, uuid);
    setIsSelfLiked(true);
  };

  const tryUnlikePost = async () => {
    await deletePostLike(sessionToken, uuid);
    onUnlike();
    setIsSelfLiked(false);
  };

  useEffect(() => {
    if (element && compact) {
      setOverflows(element.scrollHeight > element.clientHeight);
    }
  }, [element]);

  const authorDisplayName = author.name ? author.name : "";
  const profilePic = author.profile_pic ? `${imageBaseUrl}/${author.profile_pic}` : '';

  const onClickExpand = () => {
    if (compact) {
      setIsExpanded(!isExpanded);
    }
  };

  const copyLinkToClipboard = (url) => {
    navigator.clipboard.writeText(url);
    setShowCopiedMessage(true);
    setTimeout(() => {
      setShowCopiedMessage(false);
    }, 2000);
  };

  return (
    <Card sx={{ p: 1, mb: 1 }}>
      { isLoggedIn && loggedInUser.uuid === author.uuid && (
        <PostMenu
          handleDelete={() => setIsDialogOpen(true)}
          handleEdit={() => navigate(`/p-edit/${uuid}`)}
        />
      )}
      <Avatar
        alt={authorDisplayName}
        src={profilePic}
        style={{ float: "left", marginRight: 10 }}
        sx={{ width: 56, height: 56 }}
      />
      <div style={{width: "90%", paddingLeft: 72}}>
        <TitleSlug
          name={author.name}
          username={author.username}
          created={created}
          uuid={uuid}
        />
        <Typography variant="subtitle1">
          {timeAgo.format(created)}
        </Typography>
        <div onClick={onClickExpand}>
          <div
            style={{
              maxHeight: isExpanded ? "inherit" : 190,
              cursor: overflows ? "pointer" : "auto",
            }}
            className="post"
            ref={(el) => setElement(el)}
          >
            <ReactMarkdown>
              {text}
            </ReactMarkdown>
          </div>
          <div style={{width: "100%", textAlign: "center"}}>
            { !isExpanded && overflows && (
              <IconButton>
                <ExpandMoreIcon cursor="pointer" />
              </IconButton>
            )}
            { isExpanded && overflows && (
              <IconButton>
                <ExpandLessIcon cursor="pointer" />
              </IconButton>
            )}
          </div>
        </div>
        { share && (
          <Paper
            sx={{p: 1, mb: 1,}}
            elevation={2}
            variant="outlined"
          >
            <Avatar
              alt={share.user.name}
              src={share.user.profile_pic ? `${imageBaseUrl}/${share.user.profile_pic}` : ''}
              style={{float: "left", marginRight: 10}}
            />
            <TitleSlug
              name={share.user.name}
              username={share.user.username}
              created={new Date(share.created_at)}
            />
            <Typography variant="subtitle1">
              {timeAgo.format(created)}
            </Typography>
            <div className="post">
              <ReactMarkdown>
                {share.text}
              </ReactMarkdown>
            </div>
          </Paper>
        )}
        <div>
          {images && images.map((i) => (
            <Link component={RouterLink} to={`/i/${i.uuid}`} key={i.uuid}>
              <img src={`${imageBaseUrl}/${i.s3_key}`} className="post-gallery" alt="" />
            </Link>
          ))}
        </div>
      </div>
      { isLoggedIn && (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
          { showReply && (
            <IconButton aria-label="reply" component={RouterLink} to={`/p/${uuid}`}>
              <ChatBubbleOutlineIcon fontSize="small" />
            </IconButton>
          )}
          { !isSelfLiked && (
            <IconButton aria-label="like" onClick={tryLikePost}>
              <FavoriteBorderIcon fontSize="small" />
            </IconButton>
          )}
          { isSelfLiked && (
            <IconButton aria-label="unlike" onClick={tryUnlikePost}>
              <FavoriteIcon fontSize="small" />
            </IconButton>
          )}
          { showShare && (
            <IconButton aria-label="share" onClick={sharePostClick}>
              <RepeatIcon fontSize="small" />
            </IconButton>
          )}
          <IconButton aria-label="link" onClick={() => copyLinkToClipboard(`${window.location.hostname}/p/${uuid}`)}>
            <LinkIcon fontSize="small" />
            { showCopiedMessage && (
              <span style={{
                position: "absolute",
                left: 24,
                fontSize: "12pt",
                width: 120,
              }}>
                Link copied!
              </span>
            )}
          </IconButton>
        </div>
      )}
      <Dialog
        open={isDialogOpen}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title">
          Delete this post?
        </DialogTitle>
        <DialogContent>
          <DialogContentText id="alert-dialog-description">
            Are you sure? This cannot be undone!
          </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Cancel</Button>
          <Button onClick={tryDelete} autoFocus>
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </Card>
  );
}
