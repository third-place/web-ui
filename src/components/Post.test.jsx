import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Post from './Post';

test('renders correctly', () => {
  // when
  const tree = renderer
    .create(
      <BrowserRouter>
        <Post
          post={{
            uuid: "1-2-3-4",
            text: "hello world",
            created_at: new Date(),
            user: {
              uuid: "1-2-3-4",
              username: "test",
              profile_pic: "",
              name: "Tester McTesterson",
            },
            images: [],
            selfLiked: false,
          }}
        />
      </BrowserRouter>
    )
    .toJSON();

  // then
  expect(tree).toMatchSnapshot();
});
