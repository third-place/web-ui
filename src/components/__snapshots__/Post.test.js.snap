// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-qevyti-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-8h65ya-MuiAvatar-root"
    style={
      Object {
        "float": "left",
        "marginRight": 10,
      }
    }
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
      data-testid="PersonIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
      />
    </svg>
  </div>
  <div
    style={
      Object {
        "paddingLeft": 72,
        "width": "90%",
      }
    }
  >
    <h6
      className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
        href="/u/test"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        <b>
          Tester McTesterson
        </b>
        <span
          style={
            Object {
              "color": "#f17887",
              "margin": "0 20px",
            }
          }
        >
          @
          test
        </span>
        <span
          style={
            Object {
              "fontSize": "smaller",
            }
          }
        >
          just now
        </span>
      </a>
    </h6>
    <div
      onClick={[Function]}
    >
      <div
        className="post"
        style={
          Object {
            "cursor": "auto",
            "maxHeight": "inherit",
          }
        }
      >
        <p>
          hello world
        </p>
      </div>
      <div
        style={
          Object {
            "textAlign": "center",
            "width": "100%",
          }
        }
      />
    </div>
    <div />
  </div>
</div>
`;
